mlir/lib/Dialect/MIOpen/Transforms/ApplyImpl.cpp

```diff
@@ -0,0 +1,112 @@
+//===- ApplyImpl.cpp ------------------------------------------------------===//
+//
+// Copyright 2020 The MLIR Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+//
+// This pass applies target implementations to host kernel funcs.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+
+#include "mlir/Dialect/GPU/Passes.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/MIOpen/Passes.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "llvm/ADT/SmallVector.h"
+
+using namespace mlir;
+
+namespace {
+
+struct MIOpenApplyImplPass
+    : public MIOpenApplyImplPassBase<MIOpenApplyImplPass> {
+
+  void runOnOperation() override {
+    ModuleOp mod = getOperation();
+    llvm::StringRef modName = miopen::MIOpenDialect::kKernelModuleName;
+    if (mod.getName() == modName)
+      return;
+
+    if (auto miopenMod = mod.lookupSymbol<ModuleOp>(modName)) {
+      SymbolTable symbolTable(mod);
+      auto *ctx = mod.getContext();
+      OpBuilder b(ctx);
+      auto loc = mod.getLoc();
+
+      SmallVector<gpu::GPUModuleOp, 8> gpuMods;
+      miopenMod->walk([&](gpu::GPUModuleOp gpuMod) {
+        auto binaryAttr = gpuMod->getAttrOfType<StringAttr>(
+            gpu::getDefaultGpuBinaryAnnotation());
+        if (!binaryAttr) {
+          gpuMod.emitOpError() << "missing gpu.binary attribute";
+          return;
+        }
+
+        gpuMods.push_back(gpuMod);
+
+        // make global constant in TopModule
+        SmallString<128> binaryCstName(
+            SymbolTable::getSymbolName(gpuMod).getValue());
+        binaryCstName.append("_gpubin_cst");
+        // uniquify if necessary
+        while (symbolTable.lookup(binaryCstName)) {
+          binaryCstName += "0";
+        }
```

Comment:
Hmmm... Can we do better than this? `_gpubin_cast00000...000` will be a possibility?

Reply:
Yeah, there's code that goes 0,1,2,... in the gpu.printf lowering, might be worth looking at that